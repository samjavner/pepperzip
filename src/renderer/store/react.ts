import * as React from "react";
import { Commands, Dispatch, Store, Update, UseStoreOptions } from "./";

export interface Msg<M extends object, U extends Update<M>> {
    type: keyof U;
    payload?: any;
}

export function useLocalStore(log: (msg: any) => void = () => ({})) {
    return function actual<
        M extends object,
        U extends Update<M>,
        C extends Commands = {}
    >({
        init,
        update,
        createCommands,
        memo = [],
    }: UseStoreOptions<M, U, C>): Store<M, U, C> {
        const [model, reducerDispatch] = React.useReducer(
            (state: M, action: Msg<M, U>) =>
                update[action.type](state, action.payload),
            init
        );

        const vm = React.useMemo(
            () => {
                const dispatch: Dispatch<U> = {} as any;
                Object.keys(update).forEach(type => {
                    (dispatch as any)[type] = (payload: any) => {
                        const msg =
                            update[type].length === 1
                                ? { type }
                                : { type, payload };
                        log(msg);
                        reducerDispatch(msg);
                    };
                });
                const commands: C = createCommands
                    ? createCommands(model, dispatch)
                    : (({} as any) as C);
                return {
                    ...model,
                    ...dispatch,
                    ...commands,
                };
            },
            [model, update, ...memo]
        );

        return vm;
    };
}
