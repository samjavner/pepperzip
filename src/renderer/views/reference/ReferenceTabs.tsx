import { faBook, faHeartbeat } from "@fortawesome/free-solid-svg-icons";
import * as React from "react";
import { connect } from "react-redux";
import { bindActionCreators, Dispatch } from "redux";
import { Navtab, Navtabs } from "../../components/Navtabs";
import { GlobalState } from "../../core/model";
import * as navigation from "../../core/navigation";

interface ReferenceTabsProps {
    activePage: navigation.ReferencePage;
    onDictionaryClick: () => void;
    onNutritionClick: () => void;
}

export const ReferenceTabs: React.SFC<ReferenceTabsProps> = props => (
    <Navtabs>
        <Navtab
            isActive={props.activePage === "dictionary"}
            icon={faBook}
            text="Dictionary"
            onClick={props.onDictionaryClick}
        />
        <Navtab
            isActive={props.activePage === "nutrition"}
            icon={faHeartbeat}
            text="Nutrition"
            onClick={props.onNutritionClick}
        />
    </Navtabs>
);

const mapStateToProps = (state: GlobalState) => ({
    activePage: navigation.selectors.getActiveReferencePage(state),
});

const mapDispatchToProps = (dispatch: Dispatch) =>
    bindActionCreators(
        {
            onDictionaryClick: () =>
                navigation.actions.setActiveReferencePage("dictionary"),
            onNutritionClick: () =>
                navigation.actions.setActiveReferencePage("nutrition"),
        },
        dispatch
    );

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(ReferenceTabs);
