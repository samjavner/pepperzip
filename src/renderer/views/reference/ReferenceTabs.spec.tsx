import * as React from "react";
import * as renderer from "react-test-renderer";
import { ReferenceTabs } from "./ReferenceTabs";

describe("ReferenceTabs", () => {
    const onDictionaryClick = () => ({} as any);
    const onNutritionClick = () => ({} as any);

    it("should render with the 'dictionary' tab active when the 'dictionary' page is active", () => {
        const component = renderer.create(
            <ReferenceTabs
                activePage="dictionary"
                onDictionaryClick={onDictionaryClick}
                onNutritionClick={onNutritionClick}
            />
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render with the 'nutrition' tab active when the 'nutrition' page is active", () => {
        const component = renderer.create(
            <ReferenceTabs
                activePage="nutrition"
                onDictionaryClick={onDictionaryClick}
                onNutritionClick={onNutritionClick}
            />
        );
        const tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should handle click on 'dictionary' tab", () => {
        const component = renderer.create(
            <ReferenceTabs
                activePage="dictionary"
                onDictionaryClick={onDictionaryClick}
                onNutritionClick={onNutritionClick}
            />
        );
        const button = component.root.find(
            node =>
                node.props.className &&
                node.props.className.indexOf("pz-navtabs-button--active") !== -1
        );
        expect(button.props.onClick).toBe(onDictionaryClick);
    });

    it("should handle click on 'nutrition' tab", () => {
        const component = renderer.create(
            <ReferenceTabs
                activePage="nutrition"
                onDictionaryClick={onDictionaryClick}
                onNutritionClick={onNutritionClick}
            />
        );
        const button = component.root.find(
            node =>
                node.props.className &&
                node.props.className.indexOf("pz-navtabs-button--active") !== -1
        );
        expect(button.props.onClick).toBe(onNutritionClick);
    });
});
