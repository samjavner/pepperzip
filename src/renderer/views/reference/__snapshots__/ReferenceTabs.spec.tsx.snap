// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReferenceTabs should render with the 'dictionary' tab active when the 'dictionary' page is active 1`] = `
Array [
  <div
    className="pz-navtabs"
  >
    <div
      className="bp3-button-group bp3-minimal"
    >
      <button
        className="bp3-button pz-navtabs-button pz-navtabs-button--active"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        type="button"
      >
        <span
          className="bp3-icon bp3-icon-book"
          icon="book"
        >
          <svg
            data-icon="book"
            height={16}
            viewBox="0 0 16 16"
            width={16}
          >
            <desc>
              book
            </desc>
            <path
              d="M2 1v14c0 .55.45 1 1 1h1V0H3c-.55 0-1 .45-1 1zm11-1h-1v7l-2-2-2 2V0H5v16h8c.55 0 1-.45 1-1V1c0-.55-.45-1-1-1z"
              fillRule="evenodd"
            />
          </svg>
        </span>
        <span
          className="bp3-button-text"
        >
          Dictionary
        </span>
      </button>
      <button
        className="bp3-button pz-navtabs-button"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        type="button"
      >
        <span
          className="bp3-icon bp3-icon-pulse"
          icon="pulse"
        >
          <svg
            data-icon="pulse"
            height={16}
            viewBox="0 0 16 16"
            width={16}
          >
            <desc>
              pulse
            </desc>
            <path
              d="M15 8h-1.46l-1.7-2.55-.02.01A.984.984 0 0 0 11 5c-.43 0-.79.27-.93.65h-.01l-1.69 4.51-1.38-8.32h-.02A.989.989 0 0 0 6 1c-.41 0-.77.25-.92.61L2.34 8H1c-.55 0-1 .45-1 1s.45 1 1 1h2c.41 0 .77-.25.92-.61l1.65-3.86 1.44 8.63h.02c.08.47.47.84.97.84.43 0 .79-.27.93-.65h.01l2.31-6.17.92 1.38.02-.01c.17.26.46.45.81.45h2c.55 0 1-.45 1-1s-.45-1-1-1z"
              fillRule="evenodd"
            />
          </svg>
        </span>
        <span
          className="bp3-button-text"
        >
          Nutrition
        </span>
      </button>
    </div>
  </div>,
  <div
    className="pz-navtabs__border"
  />,
]
`;

exports[`ReferenceTabs should render with the 'nutrition' tab active when the 'nutrition' page is active 1`] = `
Array [
  <div
    className="pz-navtabs"
  >
    <div
      className="bp3-button-group bp3-minimal"
    >
      <button
        className="bp3-button pz-navtabs-button"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        type="button"
      >
        <span
          className="bp3-icon bp3-icon-book"
          icon="book"
        >
          <svg
            data-icon="book"
            height={16}
            viewBox="0 0 16 16"
            width={16}
          >
            <desc>
              book
            </desc>
            <path
              d="M2 1v14c0 .55.45 1 1 1h1V0H3c-.55 0-1 .45-1 1zm11-1h-1v7l-2-2-2 2V0H5v16h8c.55 0 1-.45 1-1V1c0-.55-.45-1-1-1z"
              fillRule="evenodd"
            />
          </svg>
        </span>
        <span
          className="bp3-button-text"
        >
          Dictionary
        </span>
      </button>
      <button
        className="bp3-button pz-navtabs-button pz-navtabs-button--active"
        onClick={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        type="button"
      >
        <span
          className="bp3-icon bp3-icon-pulse"
          icon="pulse"
        >
          <svg
            data-icon="pulse"
            height={16}
            viewBox="0 0 16 16"
            width={16}
          >
            <desc>
              pulse
            </desc>
            <path
              d="M15 8h-1.46l-1.7-2.55-.02.01A.984.984 0 0 0 11 5c-.43 0-.79.27-.93.65h-.01l-1.69 4.51-1.38-8.32h-.02A.989.989 0 0 0 6 1c-.41 0-.77.25-.92.61L2.34 8H1c-.55 0-1 .45-1 1s.45 1 1 1h2c.41 0 .77-.25.92-.61l1.65-3.86 1.44 8.63h.02c.08.47.47.84.97.84.43 0 .79-.27.93-.65h.01l2.31-6.17.92 1.38.02-.01c.17.26.46.45.81.45h2c.55 0 1-.45 1-1s-.45-1-1-1z"
              fillRule="evenodd"
            />
          </svg>
        </span>
        <span
          className="bp3-button-text"
        >
          Nutrition
        </span>
      </button>
    </div>
  </div>,
  <div
    className="pz-navtabs__border"
  />,
]
`;
